x/programs/rust/examples/automated-market-maker/src/lib.rs

```diff
@@ -1,261 +1,249 @@
+    token::transfer_from(&token_out, *program.account(), context.actor(), amount_out);
 
-    const PROGRAM_PATH: &str = env!("PROGRAM_PATH");
+    // update the allowance for token_in
+    assert!(token::approve(
+        &token_in,
+        *program.account(),
+        reserve_token_in + amount
+    ));
 
-    #[test]
-    fn init_state() {
-        let mut simulator = simulator::ClientBuilder::new().try_build().unwrap();
+    amount_out
+}
 
-        let program_id = simulator
-            .run_step(&Step::create_program(PROGRAM_PATH))
-            .unwrap()
-            .id;
+/// Adds 'amount_x' of token_x and 'amount_y' of token_y to the pool.
+/// The ratio of the tokens must be the same as the ratio of the tokens
+/// in the pool, otherwise the function will fail.
+/// Both tokens must be approved by the actor before calling this function
+/// Returns the amount of LP shares minted
+#[public]
+pub fn add_liquidity(context: Context<StateKeys>, amount_x: u64, amount_y: u64) -> u64 {
+    let program = context.program();
+    let (token_x, token_y) = external_token_contracts(&context);
+    let lp_token = external_liquidity_token(&context);
+
+    // calculate the amount of tokens in the pool
+    let (reserve_x, reserve_y) = reserves(&token_x, &token_y);
+
+    // ensure the proper ratio
+    assert_eq!(
+        reserve_x * amount_y,
+        reserve_y * amount_x,
+        "invalid token amounts provided"
+    );
 
-        let test_context = TestContext::from(program_id);
+    // transfer tokens from the actor to the pool
+    token::transfer_from(&token_x, context.actor(), *program.account(), amount_x);
+    token::transfer_from(&token_y, context.actor(), *program.account(), amount_y);
 
-        let resp_err = simulator
-            .run_step(&Step {
-                endpoint: Endpoint::Execute,
-                method: "remove_liquidity".to_string(),
-                max_units: u64::MAX,
-                params: vec![test_context.clone().into(), 100000u64.into()],
-            })
-            .unwrap()
-            .result
-            .response::<(u64, u64)>()
-            .unwrap_err();
-
-        let StepResponseError::ExternalCall(call_err) = resp_err else {
-            panic!("wrong error returned");
-        };
-
-        assert!(matches!(call_err, ExternalCallError::CallPanicked));
-
-        let resp_err = simulator
-            .run_step(&Step {
-                endpoint: Endpoint::Execute,
-                method: "swap".to_string(),
-                max_units: u64::MAX,
-                params: vec![test_context.into(), 100000u64.into(), true.into()],
-            })
-            .unwrap()
-            .result
-            .response::<u64>()
-            .unwrap_err();
+    // calculate the amount of shares to mint
+    let total_shares = token::total_supply(&lp_token);
+    let shares = if total_shares == 0 {
+        // if the pool is empty, mint the shares
+        math::sqrt(amount_x * amount_y)
```

Comment:
Any reason you don't mint and lock the `MINIMUM_LIQUIDITY` here ? They do it in the [`UniswapV2`](https://github.com/Uniswap/v2-core/blob/ee547b17853e71ed4e0101ccfd52e70d5acded58/contracts/UniswapV2Pair.sol#L121C30-L121C47) to prevent from inflation attacks. ref https://dapp.org.uk/reports/uniswapv2.html#orgdeb0867

Reply:
interesting attack but I don't think it'll add much to the example for devs